### ISSUE TITLE
Thoughts on target-typed new expressions?

### ISSUE THREAD

**Issue by shashank291** [hassanhabib/CSharpCodingStandard#11]
This is probably too new to form any sort of recommendation, but what are your thoughts on the C# 9 syntax for target-typed new expressions like "X object = new();"? Once you get over the initial unfamiliarity of them, I think they avoid the duplication that is needed while writing "X object = new X()" while also preserving the class name on the left side of the expression, which is better for readability. This seems easier and faster to scan than "var object = new X()".

**Comment by hassanhabib** [comment]
@shashank291 I thought about this for a while and here's my recommendation:

The `new ();` expression might cause engineers to forget what was the type they were working with before they moved on to the next line.

```csharp
var student = new Student();
```
Seems to still have superiority over the `new ()` especially when it comes to passing in variables. If we do `new (name, description)` I don't know what I'm passing name and description for. It's not self-explanatory. Now I have to go back to the beginning of the line to find out what was the type that I wanted to pass a `name` and `description` for.

I recommend against using `new ();`

**Comment by egil** [comment]
I like `new();`, but tend to only use it when the thing being new'ed up is obvious on the same line.

```c#
// In fields it nice, especially with long generic types like dictionaries.
private readonly Student student = new();

// Same for property initializers
public Dictionary<StudentId, Student> Students { get; ) = new();

// For one-liner methods, e.g. factories
public static Student Create(string name)
  => new(name);
```

For long methods, where `new()` is used with an return statement and inferred based on the method signature, I am much more careful. Here I follow the rule that it should be very close to the return type declaration so ones eyes does not have to travel far.

Used together with `var` is not an option, so that is something we don't have to worry about, e.g. this does not compile:

`var student = new();`

In summary, `new()` is not much different from `var`, which we're using without problems. 

It enables us to write code where we only have to specify the type of a field/property in one place on a line, at one side of the `=` operator, and can make code less verbose without losing clarity in my experience.
